{"cells":[{"cell_type":"markdown","metadata":{"id":"TBQGIQZJcruM"},"source":["# **Xử Lý Ngoại Lệ Trong Python**"]},{"cell_type":"markdown","metadata":{"id":"wx7tCrvbcy6U"},"source":["## Mục lục\n","\n","1. Ngoại lệ trong Python\n","2. Các loại ngoại lệ khác nhau trong Python\n","3. Sự khác biệt giữa Lỗi Cú Pháp và Ngoại Lệ\n","4. Bắt và Xử lý Ngoại lệ bằng khối lệnh Try Except<br>\n","  4.1 Bắt Ngoại lệ Cụ thể<br>\n","  4.2 Thử với Mệnh đề Else<br>\n","  4.3 Từ khóa finally trong Python<br>\n","5. Xây dựng exception riêng trong python với câu lệnh Raise\n","\n","*End*\n"]},{"cell_type":"markdown","metadata":{"id":"aldL5rvn05PQ"},"source":["#### **Warning:**  Các ví dụ trong bài sẽ sử dụng một số mã lỗi để mô phỏng các tình huống xử lý ngoại lệ trong Python. Đừng lo lắng nếu gặp phải các thông báo lỗi - đó chỉ là một phần của quá trình học và thử nghiệm. Hope everyone enjoys!"]},{"cell_type":"markdown","metadata":{"id":"W-aK06bIc2UJ"},"source":["## 1. Ngoại lệ trong Python:\n","\n","Trong Python có thể có hai loại lỗi:\n","- Lỗi cú pháp (Syntax errors).\n","- Ngoại lệ (Exceptions).\n","\n","Lỗi (Errors) là vấn đề gặp phải trong chương trình, gây ra việc dừng thực thi của chương trình. Mặt khác, ngoại lệ xảy ra khi có các sự kiện nội tại làm thay đổi luồng bình thường của chương trình.\n","\n"]},{"cell_type":"markdown","metadata":{"id":"KxC2EwwMfC6H"},"source":["Ví dụ:"]},{"cell_type":"markdown","metadata":{"id":"h3UfwGw9fLUQ"},"source":["- Lỗi cú pháp (Syntax error):"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":106},"executionInfo":{"elapsed":14,"status":"error","timestamp":1711586332877,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"UdZ0TQB1fDqq","outputId":"0275d3c0-973d-44ad-aef4-2fd12a23c47d"},"outputs":[{"ename":"SyntaxError","evalue":"expected ':' (<ipython-input-1-da876d19b495>, line 2)","output_type":"error","traceback":["\u001b[0;36m  File \u001b[0;32m\"<ipython-input-1-da876d19b495>\"\u001b[0;36m, line \u001b[0;32m2\u001b[0m\n\u001b[0;31m    if age > 18\u001b[0m\n\u001b[0m               ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m expected ':'\n"]}],"source":["# Lỗi cú pháp: Thiếu dấu hai chấm (:) sau câu lệnh if\n","if age > 18\n","    print(\"You are an adult\")\n"]},{"cell_type":"markdown","metadata":{"id":"IGWkvg55fU6v"},"source":["- Ngoại lệ (Exception):"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":176},"executionInfo":{"elapsed":377,"status":"error","timestamp":1711586337908,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"N-51JMCtbQBw","outputId":"450d431e-4788-47bc-d7ae-8eae89cbb154"},"outputs":[{"ename":"ZeroDivisionError","evalue":"division by zero","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-2-d707037ccc30>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Ngoại lệ: Chia một số cho 0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m10\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"]}],"source":["# Ngoại lệ: Chia một số cho 0\n","result = 10 / 0\n","print(result)\n"]},{"cell_type":"markdown","metadata":{"id":"cdyPXoxVfwqh"},"source":["## 2. Các loại ngoại lệ khác nhau trong Python:\n","\n","Trong Python, có một số ngoại lệ Python tích hợp có thể xảy ra khi có lỗi xảy ra trong quá trình thực thi chương trình. Dưới đây là một số loại ngoại lệ phổ biến nhất trong Python:\n","\n","- **SyntaxError**: Ngoại lệ này được tạo ra khi trình thông dịch gặp một lỗi cú pháp trong mã, như từ khóa viết sai, thiếu dấu hai chấm hoặc dấu ngoặc đơn không cân đối.\n","- **TypeError**: Ngoại lệ này được tạo ra khi một phép toán hoặc hàm được áp dụng cho một đối tượng có kiểu sai, chẳng hạn như cộng một chuỗi với một số nguyên.\n","- **NameError**: Ngoại lệ này được tạo ra khi tên biến hoặc hàm không được tìm thấy trong phạm vi hiện tại.\n","- **IndexError**: Ngoại lệ này được tạo ra khi một chỉ số nằm ngoài phạm vi của một danh sách, tuple hoặc các loại chuỗi khác.\n","- **KeyError**: Ngoại lệ này được tạo ra khi không tìm thấy một khóa trong một từ điển.\n","- **ValueError**: Ngoại lệ này được tạo ra khi một hàm hoặc phương thức được gọi với một đối số hoặc đầu vào không hợp lệ, chẳng hạn như cố gắng chuyển đổi một chuỗi thành một số nguyên khi chuỗi không đại diện cho một số nguyên hợp lệ.\n","- **AttributeError**: Ngoại lệ này được tạo ra khi một thuộc tính hoặc phương thức không được tìm thấy trên một đối tượng, chẳng hạn như cố gắng truy cập một thuộc tính không tồn tại của một thể hiện lớp.\n","- **IOError**: Ngoại lệ này được tạo ra khi một hoạt động I/O, chẳng hạn như đọc hoặc ghi một tệp, thất bại do một lỗi nhập/ra.\n","- **ZeroDivisionError**: Ngoại lệ này được tạo ra khi cố gắng chia một số cho không.\n","- **ImportError**: Ngoại lệ này được tạo ra khi một câu lệnh nhập không thể tìm thấy hoặc tải một module."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":106},"executionInfo":{"elapsed":316,"status":"error","timestamp":1711586974792,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"mp6Rv7wjg7gn","outputId":"e4cd1aa5-d4c5-47aa-b7c3-af1c3d54799e"},"outputs":[{"ename":"SyntaxError","evalue":"unterminated string literal (detected at line 2) (<ipython-input-3-21385a0a5fd3>, line 2)","output_type":"error","traceback":["\u001b[0;36m  File \u001b[0;32m\"<ipython-input-3-21385a0a5fd3>\"\u001b[0;36m, line \u001b[0;32m2\u001b[0m\n\u001b[0;31m    print(\"Hello world'\u001b[0m\n\u001b[0m          ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unterminated string literal (detected at line 2)\n"]}],"source":["# SyntaxError: Thiếu dấu ngoặc đơn\n","print(\"Hello world'"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":158},"executionInfo":{"elapsed":293,"status":"error","timestamp":1711586975467,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"BGouyjAphEu-","outputId":"6eb4d965-2d25-400f-c37f-5b4a81f85a68"},"outputs":[{"ename":"TypeError","evalue":"can only concatenate str (not \"int\") to str","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-4-b825d1fb95a8>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# TypeError: Cộng một chuỗi với một số nguyên\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"Hello\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m5\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mTypeError\u001b[0m: can only concatenate str (not \"int\") to str"]}],"source":["# TypeError: Cộng một chuỗi với một số nguyên\n","result = \"Hello\" + 5"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":158},"executionInfo":{"elapsed":291,"status":"error","timestamp":1711586978506,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"g3r6XAF8hJJd","outputId":"63beb3fb-ea59-484c-ed67-753949b978e1"},"outputs":[{"ename":"NameError","evalue":"name 'x' is not defined","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-5-8cf2482c47a7>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# NameError: Lỗi tên biến không tồn tại\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mNameError\u001b[0m: name 'x' is not defined"]}],"source":["# NameError: Lỗi tên biến không tồn tại\n","print(x)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":176},"executionInfo":{"elapsed":296,"status":"error","timestamp":1711586980587,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"0ZljoipKhO8U","outputId":"a28ee723-8835-4aca-80af-f346ccf64548"},"outputs":[{"ename":"IndexError","evalue":"list index out of range","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)","\u001b[0;32m<ipython-input-6-780925115db7>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# IndexError: Lỗi chỉ số vượt quá phạm vi của danh sách\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mmy_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmy_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mIndexError\u001b[0m: list index out of range"]}],"source":["# IndexError: Lỗi chỉ số vượt quá phạm vi của danh sách\n","my_list = [1, 2, 3]\n","print(my_list[5])"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":176},"executionInfo":{"elapsed":7,"status":"error","timestamp":1711586982710,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"tCYAkfw3hu3P","outputId":"da40bd8d-5fe1-4092-db52-0df7dbd211ad"},"outputs":[{"ename":"KeyError","evalue":"'c'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)","\u001b[0;32m<ipython-input-7-30ed373b1688>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# KeyError: Lỗi không tìm thấy khóa trong từ điển\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mmy_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"a\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"b\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmy_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"c\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mKeyError\u001b[0m: 'c'"]}],"source":["# KeyError: Lỗi không tìm thấy khóa trong từ điển\n","my_dict = {\"a\": 1, \"b\": 2}\n","print(my_dict[\"c\"])"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":158},"executionInfo":{"elapsed":295,"status":"error","timestamp":1711587056952,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"av6TnUUChTXs","outputId":"c39327f7-c694-4fea-ba93-b140a91aeeb8"},"outputs":[{"ename":"ValueError","evalue":"invalid literal for int() with base 10: 'abc'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)","\u001b[0;32m<ipython-input-13-d47cfa8402a0>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# ValueError: Lỗi giá trị không hợp lệ khi chuyển đổi kiểu\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mnum\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"abc\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: 'abc'"]}],"source":["# ValueError: Lỗi giá trị không hợp lệ khi chuyển đổi kiểu\n","num = int(\"abc\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":176},"executionInfo":{"elapsed":489,"status":"error","timestamp":1711586986943,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"TCoQktbshYrd","outputId":"779bb4ca-abe1-416e-cfde-6c88ec0cd748"},"outputs":[{"ename":"AttributeError","evalue":"'list' object has no attribute 'size'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)","\u001b[0;32m<ipython-input-9-cf6e0c0b305f>\u001b[0m in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mmy_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mmy_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmy_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mAttributeError\u001b[0m: 'list' object has no attribute 'size'"]}],"source":["# AttributeError: Lỗi không tìm thấy thuộc tính của đối tượng\n","my_list = [1, 2, 3]\n","my_list.append(4)\n","print(my_list.size)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":176},"executionInfo":{"elapsed":9,"status":"error","timestamp":1711586989192,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"WV0vm3FSh4JL","outputId":"6336ab73-eb90-40ed-da55-141e4f417b6c"},"outputs":[{"ename":"FileNotFoundError","evalue":"[Errno 2] No such file or directory: 'nonexistent.txt'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-10-ca5650cc6fd4>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# IOError: Lỗi khi đọc một tệp không tồn tại\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"nonexistent.txt\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mcontent\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'nonexistent.txt'"]}],"source":["# IOError: Lỗi khi đọc một tệp không tồn tại\n","with open(\"nonexistent.txt\", \"r\") as file:\n","    content = file.read()"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":158},"executionInfo":{"elapsed":9,"status":"error","timestamp":1711586991575,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"3P31jRh2h_3k","outputId":"470864b1-f46b-469f-b1bc-89c1c38dc802"},"outputs":[{"ename":"ZeroDivisionError","evalue":"division by zero","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-11-c9d4128f86dd>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# ZeroDivisionError: Lỗi chia một số cho không\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m10\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"]}],"source":["# ZeroDivisionError: Lỗi chia một số cho không\n","result = 10 / 0"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":331},"executionInfo":{"elapsed":8,"status":"error","timestamp":1711586993689,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"y8j7KFS5iD47","outputId":"1229f9c5-c106-463a-cdf6-79ec3b795efc"},"outputs":[{"ename":"ModuleNotFoundError","evalue":"No module named 'non_existent_module'","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)","\u001b[0;32m<ipython-input-12-dfbe78da6cc3>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# ImportError: Lỗi khi import một module không tồn tại\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mnon_existent_module\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'non_existent_module'","","\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"]}],"source":["# ImportError: Lỗi khi import một module không tồn tại\n","import non_existent_module"]},{"cell_type":"markdown","metadata":{"id":"R1f3ChpUgqwN"},"source":["## 3. Sự khác biệt giữa Lỗi Cú Pháp và Ngoại Lệ"]},{"cell_type":"markdown","metadata":{"id":"cJ-5tGxRjE7V"},"source":["**Lỗi Cú Pháp (Syntax errors)**: Như tên gọi, lỗi này xảy ra do sai cú pháp trong mã. Nó dẫn đến việc chương trình bị dừng.\n","\n","Ví dụ:\n","\n","Trong đoạn mã dưới đây có 2 lỗi cú pháp. Câu lệnh 'if' nên được theo sau bởi một dấu hai chấm (:), và câu lệnh 'print' nên được thụt vào để nằm trong khối 'if'."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":106},"executionInfo":{"elapsed":296,"status":"error","timestamp":1711588705679,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"abVLKdLIjGSU","outputId":"1efa3e67-1456-48fd-9f57-3cc25022662d"},"outputs":[{"ename":"SyntaxError","evalue":"expected ':' (<ipython-input-14-f160e68b59a3>, line 2)","output_type":"error","traceback":["\u001b[0;36m  File \u001b[0;32m\"<ipython-input-14-f160e68b59a3>\"\u001b[0;36m, line \u001b[0;32m2\u001b[0m\n\u001b[0;31m    if(ielts >= 6.5)\u001b[0m\n\u001b[0m                    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m expected ':'\n"]}],"source":["ielts = 6.5\n","if(ielts >= 6.5)\n","print(\"Chúc mừng, bạn đã đủ điểm đi du học!\")"]},{"cell_type":"markdown","metadata":{"id":"QYngzBRxjHjx"},"source":["**Ngoại Lệ (Exceptions)**: Ngoại lệ được kích hoạt khi chương trình có cú pháp chính xác, nhưng mã lại gây ra một lỗi. Lỗi này không dừng việc thực thi của chương trình, tuy nhiên, nó thay đổi luồng bình thường của chương trình.\n","\n","Ví dụ:\n","\n","Trong đoạn mã này, vì đang chia 'diem_so' cho số 0 nên gây ra một lỗi được gọi là 'ZeroDivisionError'."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":176},"executionInfo":{"elapsed":317,"status":"error","timestamp":1711588708780,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"IqsJeIfKjKaT","outputId":"4afdf023-9230-42af-bf6a-034fb8dc17df"},"outputs":[{"ename":"ZeroDivisionError","evalue":"division by zero","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-15-8bb9d9ecd9d6>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mdiem_so\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdiem_so\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"]}],"source":["diem_so = 10\n","a = diem_so / 0\n","print(a)"]},{"cell_type":"markdown","metadata":{"id":"TZjxHPPXjMDD"},"source":["Lưu ý: Ngoại lệ là lớp cơ bản cho tất cả các ngoại lệ trong Python. Bạn có thể kiểm tra cấu trúc ngoại lệ tại [đây](https://docs.python.org/2/library/exceptions.html#exception-hierarchy).\n"]},{"cell_type":"markdown","metadata":{"id":"7OCD0YbEje34"},"source":["\n","## 4. Bắt và Xử lý Ngoại lệ bằng khối lệnh Try Except"]},{"cell_type":"markdown","metadata":{"id":"fl01694boEPy"},"source":["Khối lệnh Try-Except được sử dụng để bắt và xử lý các ngoại lệ trong Python. Các câu lệnh có thể gây ra ngoại lệ được giữ bên trong mệnh đề try và các câu lệnh xử lý ngoại lệ được viết bên trong mệnh đề except."]},{"cell_type":"markdown","metadata":{"id":"47dBvV1kopWa"},"source":["Trong ví dụ sau, chúng ta có một chương trình Python cố gắng mở một tệp và đọc nội dung từ nó. Tuy nhiên, nếu tệp không tồn tại, một ngoại lệ IOError sẽ xảy ra. Chúng ta sẽ sử dụng câu lệnh try và except để bắt và xử lý ngoại lệ này."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":517,"status":"ok","timestamp":1711588751821,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"hA5pCaFKoqtz","outputId":"8efa12ff-4a44-430d-fd4a-e393e824097c"},"outputs":[{"name":"stdout","output_type":"stream","text":["Không thể mở hoặc đọc tệp.\n"]}],"source":["try:\n","    with open(\"file.txt\", \"r\") as file:\n","        content = file.read()\n","        print(\"Nội dung của tệp là:\", content)\n","except IOError:\n","    print(\"Không thể mở hoặc đọc tệp.\")\n"]},{"cell_type":"markdown","metadata":{"id":"QpbQl6RPoqBH"},"source":["Trong đoạn mã trên, câu lệnh mở tệp và đọc nội dung của nó được đặt trong khối try. Nếu tệp \"file.txt\" không tồn tại hoặc không thể đọc được vì một lý do nào đó, một ngoại lệ IOError sẽ xảy ra. Câu lệnh except sẽ bắt ngoại lệ này và in ra thông báo \"Không thể mở hoặc đọc tệp.\". Điều này giúp chương trình chạy một cách trơn tru, ngay cả khi có lỗi xảy ra khi đọc tệp."]},{"cell_type":"markdown","metadata":{"id":"8jmZuWiSpKW4"},"source":["###4.1 Bắt Ngoại lệ Cụ thể\n","\n","Một câu lệnh Try có thể có nhiều hơn một mệnh đề except, để chỉ định các trình xử lý cho các ngoại lệ khác nhau. Vui lòng lưu ý rằng tối đa một trình xử lý sẽ được thực thi. Ví dụ, chúng ta có thể thêm IndexError vào mã trên. Cú pháp tổng quát cho việc thêm các ngoại lệ cụ thể là:\n","```\n","try:\n","    # Các câu lệnh\n","except IndexError:\n","    # Các câu lệnh\n","except ValueError:\n","    # Các câu lệnh\n","```\n"]},{"cell_type":"markdown","metadata":{"id":"LkB2gtpjp0l0"},"source":["- Trong ví dụ này, chúng ta có một hàm divide(a, b) nhận hai số và cố gắng thực hiện phép chia. Trong khối try, chúng ta thực hiện phép chia và in ra kết quả. Nếu có lỗi xảy ra, chúng ta sử dụng các khối except để bắt và xử lý các loại lỗi cụ thể, bao gồm ZeroDivisionError khi chia cho 0, TypeError khi kiểu dữ liệu không phù hợp và một khối except chung để bắt các lỗi không được xác định."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1711589715330,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"YwhND_XSpzyS","outputId":"ef32135e-1a87-48e7-f309-84867e505a4e"},"outputs":[{"name":"stdout","output_type":"stream","text":["Kết quả của phép chia: 5.0\n","Lỗi: Không thể chia một số cho 0.\n","Lỗi: Loại dữ liệu không hợp lệ.\n"]}],"source":["def divide(a, b):\n","    try:\n","        result = a / b\n","        print(\"Kết quả của phép chia:\", result)\n","    except ZeroDivisionError:\n","        print(\"Lỗi: Không thể chia một số cho 0.\")\n","    except TypeError:\n","        print(\"Lỗi: Loại dữ liệu không hợp lệ.\")\n","    except Exception as e:\n","        print(\"Lỗi không xác định:\", e)\n","\n","# Thử các trường hợp chia số nguyên\n","divide(10, 2)  # Kết quả: 5.0\n","divide(8, 0)   # Lỗi: Không thể chia một số cho 0.\n","divide('a', 2) # Lỗi: Loại dữ liệu không hợp lệ."]},{"cell_type":"markdown","metadata":{"id":"jehPrkButL_f"},"source":["- Trong ví dụ 2, ta định nghĩa một hàm f(a) tính toán b dựa trên đầu vào a. Nếu a nhỏ hơn 4, nó cố gắng chia cho 0, gây ra một ZeroDivisionError. Ta gọi f(3) và f(5) bên trong một khối try-except. Nó xử lý ZeroDivisionError cho f(3) và in ra “ZeroDivisionError đã xảy ra và đã được xử lý.” Khối NameError không được thực thi vì không có các ngoại lệ NameError trong mã."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":311,"status":"ok","timestamp":1711589985318,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"lCUICpYXsqHH","outputId":"fc9a4f0f-d3ef-4c38-e2d1-fd971a09fad8"},"outputs":[{"name":"stdout","output_type":"stream","text":["ZeroDivisionError đã xảy ra và đã được xử lý\n"]}],"source":["def f(a):\n","    if a < 4:\n","        b = a/(a-3)\n","    print(\"Giá trị của b = \", b)\n","\n","try:\n","    f(3)\n","    f(5)\n","except ZeroDivisionError:\n","    print(\"ZeroDivisionError đã xảy ra và đã được xử lý\")\n","except NameError:\n","    print(\"NameError đã xảy ra và đã được xử lý\")\n"]},{"cell_type":"markdown","metadata":{"id":"PR27FO0Et22D"},"source":["Nếu comment dòng f(3), kết quả sẽ là:\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":7,"status":"ok","timestamp":1711590106911,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"9akW8AsCuAcF","outputId":"10fceec6-2644-4661-f180-583940a7ab01"},"outputs":[{"name":"stdout","output_type":"stream","text":["NameError đã xảy ra và đã được xử lý\n"]}],"source":["def f(a):\n","    if a < 4:\n","        b = a/(a-3)\n","    print(\"Giá trị của b = \", b)\n","\n","try:\n","    #f(3)\n","    f(5)\n","except ZeroDivisionError:\n","    print(\"ZeroDivisionError đã xảy ra và đã được xử lý\")\n","except NameError:\n","    print(\"NameError đã xảy ra và đã được xử lý\")\n"]},{"cell_type":"markdown","metadata":{"id":"egKSKYWvuHYa"},"source":["Kết quả trên là do ngay khi python cố gắng truy cập giá trị của b, NameError xảy ra."]},{"cell_type":"markdown","metadata":{"id":"wfCH-U-WuKxJ"},"source":["### 4.2 Thử với Mệnh đề Else:\n","\n","Trong Python, bạn cũng có thể sử dụng mệnh đề else trong khối try-except, mà phải xuất hiện sau tất cả các mệnh đề except. Mã nhập vào khối else chỉ được thực hiện nếu mệnh đề try không gây ra ngoại lệ."]},{"cell_type":"markdown","metadata":{"id":"ybyERHXHuvJa"},"source":["- Trong ví dụ này, hàm calculate_ratio tính toán tỉ số giữa a và b. Nếu không có lỗi chia cho 0 xảy ra, kết quả sẽ được in ra trong khối else. Nếu có lỗi chia cho 0 xảy ra, ngoại lệ ZeroDivisionError sẽ được bắt và thông báo sẽ được in ra."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":306,"status":"ok","timestamp":1711590486028,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"EQglY8YovSFW","outputId":"9cfec1f2-dff0-4fe0-8591-d8f98895838f"},"outputs":[{"name":"stdout","output_type":"stream","text":["Kết quả của phép tính: -7.0\n","Lỗi: a/b kết quả là 0\n"]}],"source":["def calculate_ratio(a, b):\n","    try:\n","        result = ((a + b) / (a - b))\n","    except ZeroDivisionError:\n","        print(\"Lỗi: a/b kết quả là 0\")\n","    else:\n","        print(\"Kết quả của phép tính:\", result)\n","\n","calculate_ratio(6, 8)\n","calculate_ratio(4, 4)\n"]},{"cell_type":"markdown","metadata":{"id":"mmL0XMtvwBjW"},"source":["### 4.3 Từ khóa finally trong Python:\n","Python cung cấp một từ khóa finally, luôn được thực thi sau các khối try và except. Khối cuối cùng luôn được thực thi sau khi khối try kết thúc bình thường hoặc sau khi khối try kết thúc do một số ngoại lệ.\n","\n","Cú pháp:\n","```\n","try:\n","    # Một số mã...\n","\n","except:\n","    # Khối tùy chọn\n","    # Xử lý ngoại lệ (nếu cần)\n","\n","else:\n","    # thực hiện nếu không có ngoại lệ\n","\n","finally:\n","    # Một số mã... (luôn được thực hiện)\n","```"]},{"cell_type":"markdown","metadata":{"id":"1Mq1rpL2wUuu"},"source":["- Trong ví dụ này, chương trình cố gắng chia 10 cho số nguyên được nhập từ người dùng. Nếu người dùng nhập số 0, một ngoại lệ ZeroDivisionError sẽ xảy ra và thông báo \"Lỗi: Không thể chia cho 0.\" sẽ được in ra. Sau đó, khối finally sẽ luôn được thực hiện và in ra \"Khối finally luôn được thực hiện\"."]},{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":5692,"status":"ok","timestamp":1711637690456,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"Ci0DZd-xwmkv","outputId":"041b3fb0-dde3-464c-f06c-26255be7945c"},"outputs":[{"name":"stdout","output_type":"stream","text":["Nhập một số nguyên: 6\n","Kết quả của phép chia: 1.6666666666666667\n","In ra nếu không có ngoại lệ\n","Khối finally luôn được thực hiện\n"]}],"source":["try:\n","    num = int(input(\"Nhập một số nguyên: \"))\n","    result = 10 / num\n","    print(\"Kết quả của phép chia:\", result)\n","except ZeroDivisionError:\n","    print(\"Lỗi: Không thể chia cho 0.\")\n","else:\n","    print(\"In ra nếu không có ngoại lệ\")\n","finally:\n","    print('Khối finally luôn được thực hiện')"]},{"cell_type":"markdown","metadata":{"id":"aBHnOvCmw8Af"},"source":["## 5. Xây dựng exception riêng trong python với câu lệnh Raise"]},{"cell_type":"markdown","metadata":{"id":"sIrauhzLxRaf"},"source":["Câu lệnh raise trong Python được sử dụng để tạo ra và ném ra một ngoại lệ cụ thể. Nó cho phép lập trình viên tạo ra các tình huống ngoại lệ theo ý muốn và thông điệp tùy chỉnh, giúp quản lý và xử lý lỗi trong chương trình một cách linh hoạt và hiệu quả. Đối số duy nhất trong câu lệnh raise chỉ định ngoại lệ sẽ được tạo ra. Đối số này phải là một phiên bản ngoại lệ hoặc một lớp ngoại lệ (một lớp kế thừa từ Exception)."]},{"cell_type":"markdown","metadata":{"id":"5RhlXEIayIVB"},"source":["- Ví dụ dưới đây cố ý tạo ra một NameError với thông điệp \"Xin chào!\" bằng cách sử dụng câu lệnh raise trong một khối try. Sau đó bắt ngoại lệ NameError, in ra \"Một ngoại lệ,\" và tiếp tục tạo ra cùng một ngoại lệ bằng cách sử dụng câu lệnh raise. Output của đoạn mã này sẽ chỉ in ra dòng \"Một ngoại lệ,\" nhưng cũng sẽ xảy ra một lỗi Runtime cuối cùng do câu lệnh raise ở dòng cuối cùng."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":228},"executionInfo":{"elapsed":305,"status":"error","timestamp":1711591322748,"user":{"displayName":"Linh Vu","userId":"05812158188680418351"},"user_tz":-420},"id":"H35QFxWFyE6t","outputId":"1f8b298c-07f2-4b10-d807-df764ffd4a82"},"outputs":[{"name":"stdout","output_type":"stream","text":["Một ngoại lệ\n"]},{"ename":"NameError","evalue":"Xin chào!","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-39-912883e7e3c2>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mNameError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Xin chào!\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;32mexcept\u001b[0m \u001b[0mNameError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m\"Một ngoại lệ\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mNameError\u001b[0m: Xin chào!"]}],"source":["try:\n","    raise NameError(\"Xin chào!\")\n","except NameError:\n","    print (\"Một ngoại lệ\")\n","    raise"]},{"cell_type":"markdown","metadata":{"id":"PAAtkLAvzy5x"},"source":["Bằng cách sử dụng raise, bạn có thể xây dựng một exception giúp xử lý ngoại lệ trong Python. Với trường hợp này, bạn có thể tạo exception của riêng mình - đó là exception được đưa ra khi vấn đề nằm bên ngoài phạm vi xảy ra của dự kiến lỗi"]},{"cell_type":"markdown","metadata":{"id":"SBjVYyYsz3IH"},"source":["- Trong ví dụ này, nếu người dùng nhập một số âm cho tuổi, câu lệnh raise sẽ ném ra một ngoại lệ ValueError với thông điệp \"Tuổi không thể là một số âm.\" sau đó, khối except sẽ được thực thi và in ra thông báo lỗi."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"cvIeCieLz1Zn"},"outputs":[],"source":["try:\n","    age = int(input(\"Nhập tuổi của bạn: \"))\n","    if age < 0:\n","        raise ValueError(\"Tuổi không thể là một số âm.\")\n","    print(\"Tuổi của bạn là:\", age)\n","except ValueError as ve:\n","    print(\"Lỗi:\", ve)\n"]},{"cell_type":"markdown","metadata":{"id":"SRixMR_20Ipm"},"source":["## *End*"]}],"metadata":{"colab":{"authorship_tag":"ABX9TyPX/BuaGb/JPL4MraZ5rnBo","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}
